### PR TITLE
Giphy gifs

### PR DESCRIPTION
**Pull request by ivolarevic**
# Description

Implemented Giphy gifs in the app.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Dev testing.

**Test Configuration**:

* Firmware version: G930FXXS7ETA1
* Hardware: Samsung galaxy s7
* SDK: Android 8, Oreo

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 1]
```diff
@@ -492,6 +534,46 @@ class ChatMessagesFragment : BaseFragment(), ServiceConnection {
         }
     }
 
+    fun handleGifClick(context: Context, urlString: String) {
+        Toast.makeText(context, getString(R.string.preparing_gifs), Toast.LENGTH_LONG).show()
+        CoroutineScope(Dispatchers.IO).launch {
+            try {
+                val url = URL(urlString)
+                val connection: HttpURLConnection = url.openConnection() as HttpURLConnection
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:542]: This connection variable should be called withContext(). The Coroutine might get blocked and lead to thread starvation. It should be destroyed with the activity/fragment it lives in.

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 2]
```diff
@@ -1464,10 +1554,21 @@ class ChatMessagesFragment : BaseFragment(), ServiceConnection {
...
+                            } ?: filesSelected.first()
+                        } else {
+                            filesSelected.first()
+                        }
+
                         uploadFiles(
                             isThumbnail = false,
-                            uri = filesSelected.first(),
+                            uri = uri,
                             localId = unsentMessage.localId!!,
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:1572]: We might try to resolve this !! now

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 3]
```diff
@@ -1316,13 +1398,13 @@ class ChatMessagesFragment : BaseFragment(), ServiceConnection {
...
-            0,
-            0,
-            roomWithUsers!!.room.roomId,
-            localId,
-            replyId
+            msgText = text,
+            mimeType = Const.JsonFields.TEXT_TYPE,
+            fileId = 0,
+            thumbId = 0,
+            roomId = roomWithUsers!!.room.roomId,
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:1405]: Try and resolve this !! here

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 4]
```diff
@@ -1502,11 +1603,11 @@ class ChatMessagesFragment : BaseFragment(), ServiceConnection {
...
             FilesHelper.uploadFile(
-                isThumbnail,
-                uri,
-                localId,
-                roomWithUsers!!.room.roomId,
-                metadata
+                isThumbnail = isThumbnail,
+                uri = uri,
+                localId = localId,
+                roomId = roomWithUsers!!.room.roomId,
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:1609]: Here too, the !!

File: app/src/main/java/com/clover/studio/spikamessenger/utils/Tools.kt [hunk 5]
```diff
@@ -702,12 +742,12 @@ object Tools {
     fun getFileType(uri: Uri): String {
         val mimeType = getFileMimeType(MainApplication.appContext, uri)
         return when {
-            mimeType?.contains(Const.JsonFields.SVG_TYPE) == true -> Const.JsonFields.FILE_TYPE
-            mimeType?.contains(Const.JsonFields.AVI_TYPE) == true -> Const.JsonFields.FILE_TYPE
-            mimeType?.contains(Const.JsonFields.MOV_TYPE) == true -> Const.JsonFields.FILE_TYPE
-            mimeType?.contains(Const.JsonFields.IMAGE_TYPE) == true -> Const.JsonFields.IMAGE_TYPE
-            mimeType?.contains(Const.JsonFields.VIDEO_TYPE) == true -> Const.JsonFields.VIDEO_TYPE
-            mimeType?.contains(Const.JsonFields.AUDIO_TYPE) == true -> Const.JsonFields.AUDIO_TYPE
+            mimeType.contains(Const.JsonFields.SVG_TYPE) -> Const.JsonFields.FILE_TYPE
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/utils/Tools.kt:745]: Should GIF types also be handled here?

File: app/src/main/java/com/clover/studio/spikamessenger/utils/Tools.kt [hunk 6]
```diff
@@ -718,9 +758,17 @@ object Tools {
...
-        return cR.getType(uri)
+    fun getFileMimeType(context: Context, uri: Uri): String {
+        val type = if (uri.toString().contains(Const.JsonFields.GIF)) {
+            Const.JsonFields.GIF_TYPE
+        } else {
+            val cR: ContentResolver = context.contentResolver
+            cR.getType(uri).toString()
+        }
+
+        Timber.d("TYPE::::::::: $type")
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/utils/Tools.kt:769]: You can remove this log

File: app/src/main/java/com/clover/studio/spikamessenger/utils/UploadDownloadManager.kt [hunk 7]
```diff
@@ -42,7 +42,12 @@ class UploadDownloadManager constructor(
         fileData: FileData,
         fileUploadListener: FileUploadListener,
     ) {
-        var mimeType = MainApplication.appContext.contentResolver.getType(fileData.fileUri)!!
+        var mimeType = MainApplication.appContext.contentResolver.getType(fileData.fileUri)
+            ?.takeIf { it.isNotEmpty() }
+            ?: if (fileData.fileUri.toString().contains(Const.JsonFields.GIF)) Const.JsonFields.IMAGE_TYPE else ""
+
+        Timber.d("MIME TYPE:::: $mimeType")
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/utils/UploadDownloadManager.kt:49]: This log can also be removed

File: app/src/main/java/com/clover/studio/spikamessenger/utils/UploadDownloadManager.kt [hunk 8]
```diff
@@ -141,22 +146,22 @@ class UploadDownloadManager constructor(
...
-                        mimeType,
-                        file.id.toLong(),
-                        0,
-                        file.type!!,
-                        messageBody
+                        path = it,
+                        mimeType = mimeType,
+                        thumbId = file.id.toLong(),
+                        fileId = 0,
+                        fileType = file.type!!,
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/utils/UploadDownloadManager.kt:153]: Try and remove !! here

File: app/src/main/java/com/clover/studio/spikamessenger/utils/UploadDownloadManager.kt [hunk 9]
```diff
@@ -141,22 +146,22 @@ class UploadDownloadManager constructor(
...
-                        mimeType,
-                        0,
-                        file.id.toLong(),
-                        file.type!!,
-                        messageBody
+                        path = it,
+                        mimeType = mimeType,
+                        thumbId = 0,
+                        fileId = file.id.toLong(),
+                        fileType = file.type!!,
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/utils/UploadDownloadManager.kt:163]: Here too !!

File: app/src/main/java/com/clover/studio/spikamessenger/utils/helpers/FilesHelper.kt [hunk 10]
```diff
@@ -93,9 +93,9 @@ object FilesHelper {
             val inputStream =
                 MainApplication.appContext.contentResolver.openInputStream(uri)
             size = Tools.copyStreamToFile(
-                inputStream!!,
-                MainApplication.appContext.contentResolver.getType(uri)!!,
-                fileName
+                inputStream = inputStream!!,
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/utils/helpers/FilesHelper.kt:96]: !! here too

**Review by MatkoTom — CHANGES_REQUESTED** [review]
Check the comments and make the required adjustments.

File: app/src/main/java/com/clover/studio/spikamessenger/utils/Tools.kt (hunk 5 above)
Comment (ivolarevic) [reply app/src/main/java/com/clover/studio/spikamessenger/utils/Tools.kt:745]: No.

**Review by MatkoTom — APPROVED** [review]
LGTM

**Pull request merged**